### PR TITLE
tx builder minting

### PR DESCRIPTION
**Pull request by iadmytro**
# Context

Tx builder extended with minting functionality
LW-9394

### CODE REVIEW

File: packages/tx-construction/src/tx-builder/PolicyBuilder.ts [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    return this.#derivationPath;
+  }
+
+  setPolicyScript(script: Cardano.NativeScript) {
+    this.#policyScript = script;
+    return this;
+  }
+
+  async getPolicyScript() {
+    if (!this.#policyScript) {
```
Comment (mirceahasegan) [inline packages/tx-construction/src/tx-builder/PolicyBuilder.ts:35]: If `derivationPath` is updated, the policy script should also be recalculated as it depends on the `keyHash`.
It would be great to have these cases covered in unit tests.
```
- if (!this.#policyScript) {
+ if (!this.#policyScript || this.#derivationPathUpdated) {

Comment (iadmytro) [reply packages/tx-construction/src/tx-builder/PolicyBuilder.ts]: updated in [c7e44c7](https://github.com/input-output-hk/cardano-js-sdk/pull/893/commits/c7e44c75007d6bbb5447b0889030b0005219a75c)

File: packages/tx-construction/src/tx-builder/types.ts [hunk 2]
```diff
@@ -223,6 +223,44 @@ export interface TxBuilder {
...
+   *  ->          .assets(tokensOutput)
+   *              .build())
+   *        .build()
+   *        .sign();
+   *      await wallet.submitTx(tx);
+   */
+  addMint(mintToken: Cardano.TokenMap): TxBuilder;
+
+  /** @param script */
+  addNativeScript(script: Cardano.NativeScript): TxBuilder;
```
[the remark this answers is not in the document]
Comment (iadmytro) [reply packages/tx-construction/src/tx-builder/types.ts]: removed in [ffbc789](https://github.com/input-output-hk/cardano-js-sdk/pull/893/commits/ffbc7896c101d6dcf61758e8d680974796a4fa6b)

File: packages/tx-construction/src/tx-builder/TokenBuilder.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+      .map((s) => s.charCodeAt(0).toString(16))
+      .join('');
+
+  addAsset(tokenName: string, quantity: bigint): TxTokenBuilder {
+    this.#assets.set(Cardano.AssetId(`${this.#policyId}${this.#toHex(tokenName)}`), quantity);
+    return this;
+  }
+
+  build(): Cardano.TokenMap {
+    return this.#assets ? this.#assets : new Map();
```
Comment (mirceahasegan) [inline packages/tx-construction/src/tx-builder/TokenBuilder.ts:24]: this.#assets is initialised to `new Map()` in constructor.
Also, please return a copy of the assets. Otherwise, the returned assets map is just a reference to the internal `#assets` of the TxTokenBuilder and updating it from the TxTokenBuilder will also update the previously returned `#assets`.
```
-     return this.#assets ? this.#assets : new Map();
+    return new Map(this.#assets);;

File: packages/tx-construction/src/tx-builder/PolicyBuilder.ts
Comment (mirceahasegan) [inline packages/tx-construction/src/tx-builder/PolicyBuilder.ts:1]: Please add unit tests. `derivationPathUpdated` cases should be covered.

File: packages/e2e/test/wallet/PersonalWallet/nft.test.ts [hunk 4]
```diff
@@ -142,47 +90,41 @@ describe('PersonalWallet.assets/nft', () => {
...
+      Cardano.AssetId(`${policyId}${assetNames[TOKEN_METADATA_2_INDEX]}`),
+      Cardano.AssetId(`${policyId}${assetNames[TOKEN_BURN_INDEX]}`)
+    ];
 
-    const finalizeProps: FinalizeTxProps = {
-      auxiliaryData,
-      tx: unsignedTx,
-      witness: { extraSigners: [policySigner], scripts: [policyScript] }
-    };
+    const tokens = new Map([
```
Comment (mirceahasegan) [inline packages/e2e/test/wallet/PersonalWallet/nft.test.ts:101]: Could you use the token builder here?

File: packages/tx-construction/src/tx-builder/types.ts [hunk 5]
```diff
@@ -223,6 +223,44 @@ export interface TxBuilder {
    * @param portfolio the CIP17 delegation portfolio to apply. Using `null` will deregister all stake keys,
    *  reclaiming the deposits.
    */
+
+  /**
+   *
+   * @param mintToken
+   * list of tokens to mint/burn should be balanced with outputs
+   * to mint tokens use tokenBuilder
+   *      const policy = await txBuilder.buildPolicy();
```
[the remark this answers is not in the document]
Comment (mirceahasegan) [reply packages/tx-construction/src/tx-builder/types.ts:269]: Added

File: packages/web-extension/src/observableWallet/util.ts [hunk 6]
```diff
@@ -38,7 +38,15 @@ export const outputBuilderProperties: RemoteApiProperties<OutputBuilder> = {
   }
 };
 
-export const txBuilderProperties: RemoteApiProperties<Omit<TxBuilder, 'customize'>> = {
+export const txBuilderProperties: RemoteApiProperties<Omit<TxBuilder, 'customize' | 'buildToken' | 'buildPolicy'>> = {
```
Comment (mirceahasegan) [inline packages/web-extension/src/observableWallet/util.ts:41]: @mkazlauskas , I excluded the policy/token build utilities from the TxBuilder remote api interface to avoid cluttering.
I believe this might have been the intent in the original PR, but it was done by excluding these methods from the TxBuilder interface. I added them in the interface, but excluded them from the exposed properties.
Do you agree?

**Comment by mirceahasegan** [comment]
There is also the `handle.test.ts` that could be migrated to the newer mint/burn api, but it currently uses well written utilities that I did not want to change.

File: packages/e2e/src/factories.ts [hunk 7]
```diff
@@ -322,7 +322,8 @@ const patchInitializeTxToRespectEpochBoundary = <T extends ObservableWallet>(
  * @returns an object containing the wallet and providers passed to it
  */
 export const getWallet = async (props: GetWalletProps) => {
-  const { env, idx, logger, name, polling, stores, customKeyParams, keyAgent, witnesser } = props;
+  const { env, idx, logger, name, polling, stores, customKeyParams, witnesser } = props;
+  const keyAgent = props.keyAgent;
```
Comment (iccicci) [inline packages/e2e/src/factories.ts:326]: Why this change?

And moreover, how could it be that the eslint rule `unicorn/consistent-destructuring` is not applied here?

**Review by iccicci — APPROVED** [review]
The benefits from this feature is clearly visible in the e2e tests! 💪

File: packages/e2e/src/factories.ts (hunk 7 above)
Comment (mirceahasegan) [reply packages/e2e/src/factories.ts:326]: Great catch!
Weird indeed. The change was due to me experimenting, and forgot to roll it back

Comment (iccicci) [reply packages/e2e/src/factories.ts:326]: Just out of curiosity, mainly about the eslint rule...
I think no actions are required here

File: packages/web-extension/src/observableWallet/util.ts (hunk 6 above)
Comment (mkazlauskas) [reply packages/web-extension/src/observableWallet/util.ts:41]: Problem with this approach is that ObservableWallet.createTxBuilder returns a TxBuilder so those methods would appear as available but just fail. It would be better to make it actually compatible with messaging by following the pattern of OutputBuilder (`build` method async).

File: packages/tx-construction/src/tx-builder/TxBuilder.ts [hunk 8]
```diff
@@ -155,14 +159,34 @@ export class GenericTxBuilder implements TxBuilder {
     });
   }
 
+  addMint(mintToken: Cardano.TokenMap): TxBuilder {
```
Comment (mkazlauskas) [inline packages/tx-construction/src/tx-builder/TxBuilder.ts:162]: Missing tests for new methods in `TxBuilder.test.ts`

**Review by mkazlauskas — CHANGES_REQUESTED** [review]
As agreed on Slack let's park this until we have a use case
